docs/rfc/block-registration.md

```diff
@@ -0,0 +1,441 @@
+		"editor": "my-block-editor-style.css",
+		"theme": "my-block-theme-style.css"
+	}
+}
+```
+
+This property is a pointer to CSS files containing the CSS used for the block in different contexts.
+
+## Backward compatibility
+
+Three properties are going to be supported for backward compatibility reasons on the client-side, and they are going to be replaced with alternative APIs in the future:
+ - `supports` - see the [block supports](https://wordpress.org/gutenberg/handbook/designers-developers/developers/block-api/block-registration/#supports-optional) documentation page for more details.
+ - `merge` - undocumented as of today. Its role is to handle merging multiple blocks into one.
+ - `getEditWrapperProps` - undocumented as well. Its role is to inject additional props to the block edit's component wrapper.
+ 
+**Example**:
+```js
+wp.blocks.registerBlockType( 'my-block/name', {
+    edit: function() {
+    	// Edit definition goes here.
+    },
+    save: function() {
+    	// Save definition goes here.
+    },
+    supports: {
+        html: false
+    }
+} );
+``` 
+
+## Internationalization
+
+Localized properties are automatically wrapped in `_x` function calls on the back end and the front end of WordPress. These translations are added as an inline script to the `wp-block-library` script handle in WordPress core or to the plugin's script handle when it defines metadata definition.
+
+WordPress string discovery automatically translates these strings using the `textdomain` value specified in the plugin header.
+
+**Example:**
+
+```json
+{
+	"title": "My block",
+	"description": "My block is fantastic",
+	"keywords": [ "fanstastic" ]
+}
+```
+
+In JavaScript, with the help of a Babel plugin, this becomes:
+
+```js
+const metadata = {
+	title: _x( 'My block', 'block title', 'my-plugin' ),
+	description: _x( 'My block is fantastic', 'block description', 'my-plugin' ),
+	keywords: [ _x( 'fanstastic', 'block keywords', 'my-plugin' ) ],
+}
+```
+
+In PHP, it is transformed at runtime to code roughly equivalent to:
+
+```php
+$metadata = array(
+	'title' => _x( 'My block', 'block title', 'my-plugin' ),
+	'description': _x( 'My block is fantastic', 'block description', 'my-plugin' ),
+	'keywords': array( _x( 'fanstastic', 'block keywords', 'my-plugin' ) ),
+);
+```
+
+Implementation should follow the existing [get_plugin_data](https://codex.wordpress.org/Function_Reference/get_plugin_data) function which parses the plugin contents to retrieve the plugin’s metadata, and it applies translations dynamically.
+
+## PHP Runtime
+
+WordPress automatically discovers all the block.json files in the plugin/core `blocks` folder and registers the corresponding block types. These block types are made available through the [block registry](https://developer.wordpress.org/reference/classes/wp_block_type_registry/) PHP class and the blocks scripts and styles are added as dependencies to the `wp-block-library` script and style handles.
```

Comment:
I'm still very confused to why `.json` files should be the source of registering blocks. Maybe I've overlooked the reasoning, but it doesn't resolve in my mind what value this brings. It feels like a step backward to have a server side registry that can't make use of the server language. 

I get that the JS client _needs_ JSON as the final format of the block registration, but can't we register blocks in PHP and the client ask for the registry from the server and have it delivered in JSON?

What am I missing?

Reply:
@jasonbahl Gutenberg is no longer just a WordPress project. There ports to Drupal and other PHP frameworks. Forcing developers to use PHP and a WordPress function (would would use globals ), it limiting. Defining in JSON, makes the definition and implementation apart. Which I like.

Reply:
Defining in JSON severely limits the usefulness. 

You can't do something like:

```
foreach( $new_block_type as $block_type ) {
  register_block_type( $block_type['name'], $block_type[ $config ]
}
```

in JSON, for example. 

There's also a LOT of things missing from Gutenberg to make it production ready, like the ability to validate input on the server when a block is edited, and a JSON file isn't going to help much with that. 

We need to be able to define things like `sanitize_callback` on Block attributes so the server can have the final say on what data is safe to be saved to the server. Right now, Gutenberg allows anything to be saved, whether it will break your application or not. 

If there's a server side registry, there will be more clients than just a JS client (CLI, REST, GraphQL, others)

And if the server can't validate input, which it can't now, then we're all in trouble. 

WordPress core features should be developed with WordPress users in mind, and if it can also work in Drupal or elsewhere, I think that's fantastic, but sacrificing the WP developer/user experience and data integrity to try and support other CMS cases doesn't help WP or the other CMS.

Reply:
I don't disagree with you. But I think gutenberg is keeping headerless in mind always. I also have to think, what problem are being trying to solve here. `sanitize_callback`, `auth_callback` and `default_values` are all something i would love to see gutenberg blocks.

Having written a number of blocks, the definition needs to be in available in core at compile time and dynamically genereated blocks wouldn't work well with that. 

I do however think that block definitions should read for the api and run through PHP, for the simple reason, they will run through filters and transaction functions.

Reply:
While I think the relationship between the two could be made clearer in this document, it's my understanding that `register_block_type` still exists and is still ultimately how the block becomes known to the server. That the non-behavioral properties of a block be defined in a static format is to simplify reusability between server and client implementations†, and to enable understanding of blocks in a broader set of use-cases by benefit of its static analyzability (e.g. the [block directory efforts](https://make.wordpress.org/meta/2019/03/08/the-block-directory-and-a-new-type-of-plugin/)).

† In both PHP and JavaScript, blocks are registered as a combination of static values from the JSON file and runtime-specific logic of their respective languages / contexts, if necessary.

Reply:
@aduth Couldn't the definition of the block come from the rest api, allowing you to register dynamic blocks purely in PHP? This would make the rest api the point of truth for the defintion of a block. There are lots of contexts like the WP CLI, where this would be useful.

Reply:
> @aduth Couldn't the definition of the block come from the rest api, allowing you to register dynamic blocks purely in PHP? This would make the rest api the point of truth for the defintion of a block. There are lots of contexts like the WP CLI, where this would be useful.

An endpoint can (and presumably will) exist in the context of a site, sure. For something like the block directory work, it needs to attain information about a block purely from the contents of a plugin, without a site or its REST API to consume from.